broadcast/broadcast.go

```
@@ -0,0 +1,283 @@
+	for _, sub := range b.subs {
+		select {
+		case sub.notifyC <- struct{}{}:
+		default:
+		}
+	}
+	return nil
+}
+
+// Subscribe adds a new subscriber starting at the current tail (next message).
+// Subscription must be ended by calling Unsubscribe to terminate forwarder goroutine and cleanup resources,
+// or after user facing channel is closed after draining the buffer when broadcast is closed.
+//
+// - Acquires subMu to add to the subs map.
+// - Reads ring.tail under ring's tail lock to set starting sequence.
+// - Spawns a goroutine to deliver messages to sub.out.
+func (b *Broadcast[T]) Subscribe() *Subscription[T] {
+	b.ring.tailMu.RLock()
+	next := b.ring.tail
+	b.ring.tailMu.RUnlock()
+
+	subID := b.nextSubID.Add(1)
+	sub := &Subscription[T]{
+		bcast:   b,
+		seq:     next, // next msg to read
+		notifyC: make(chan struct{}, 1),
+		out:     make(chan EventOrLag[T], 1),
+		done:    make(chan struct{}),
+		id:      subID,
+	}
+
+	b.subMu.Lock()
+	b.subs[sub.id] = sub
+	b.subMu.Unlock()
+	go sub.run()
+
+	return sub
+}
+
+// Close marks the broadcast as closed (single-run via CAS),
+// closes all subscribers notifyC channel(drain-mode), and clears the subs map.
+// It does not close per-sub user facing channels here;
+// each subscription’s run goroutine is responsible for exiting (draining up to the latest) and
+// will close its out channel on return.
+func (b *Broadcast[T]) Close() {
+	if !b.closed.CompareAndSwap(false, true) {
+		return
+	}
+
+	b.subMu.Lock()
+	defer b.subMu.Unlock()
+
+	for _, sub := range b.subs {
+		close(sub.notifyC) // Drain mode
+	}
+	b.subs = make(map[uint64]*Subscription[T])
+}
+
+// Unsubscribe removes this subscriber from the broadcast (if still present) and
+// closes its control channels.
+//   - single-run via CAS.
+//   - Deletes from subs under subMu.Lock, then closes done (to stop run).
+func (sub *Subscription[T]) Unsubscribe() {
+	if !sub.isClosed.CompareAndSwap(false, true) {
+		return
+	}
```

Should unsubscribing of an already closed subscription silently do nothing or return an error. I think the second might be more appropriate 🤔

---

I'm not sure, there are multiple paths were Unsubscribe is called, internally or by consumer. If we wont gonna avoid multiple unsubscribe paths, idempotent behavior is better since in such cases error would force caller to branch or log noise for something that isn't actually exceptional. Example of internal unsubscribe paths are Unsubscribing to cleanup resources and terminate connection upon unexpected error, or broadcast closes then after draining the buffer, forwarder goroutine unsubscribes and exits since no more values will be sent and received. More like canceling same context twice does not yield in error or panic since subsequent calls are no-op.

---

Ok, let's leave it as no op 👍